---
title: Versioning your Docs
description: How to turn on and use versioning
---

# Software Overview

**ESGI** is a debarcoding tool for single-cell sequencing data, consisting of two submodules: **demultiplex** and **count**. You can run these submodules separately or execute ESGI to run both steps and generate a single-cell feature count matrix. For more details, run demultiplex, count, or ESGI with the  `--help` flag. 

## Demultiplex
Demultiplexing maps sequencing reads to barcode schemes, where positional barcode patterns are used to encode experiment-specific information, like cell identities, molecular modalities, and experimental conditions. The tool handles simultaneous mapping to multiple barcode schemes, supports positional barcode patterns of varying lengths, and allows for mismatches in the barcode patterns arising from insertions, deletions, and substitutions. 

### Set up:
To demultiplex barcode sequences, you need the following input files:

| Option | Description | File Type |
| --------- | ----------- | ------ | 
| `--input`, `-i` | Single-end or forward read file | fastq(.gz) 
| `--reverse`, `-r` | Reverse read file (optional) | fastq(.gz)
| `--BarcodePatternsFile`, `-p` | Description of the barcode scheme, using bracket-enclosed sequence substrings to define the position of barcode patterns. Each bracket contains a comma separated list of possible barcode sequences for that position, and these barcodes may vary in length | (.txt)
|  `--mismatchFile`, `-m` |  A comma-separated list of integers, one for each pattern in the barcode scheme, specifying the number of mismatches allowed for in each bracket-enclosed substring | (.txt) 

Example of a barcode scheme consisting of six positional barcode patterns, including the number of mismatches allowed for per pattern. 
```
# Barcode scheme with six positional patterns
-----> <---------------------------------
[RNA][-][BC1.txt][AGCTCATC][BC2.txt][10X]
|    |  |        |                  └─── Sequence for ten random bases
|    |  |        └─── Constant barcode
|    |  └─── List of possible sequences
|    └─── Read separator for forward and reverse read
└─── Sequence of transcript

# Allowed mismatches per barcode pattern
1,0,1,0,1,0
```

Optional parameters:

| Option | Description | Default |
| --------- | ----------- | ------ | 
| `--independent`, `-d` | Treat the forward read as two separate sequences in the 5'-->3' direction. Use together with the read separator [-] in the `--BarcodePatternsFile` to indicate where one read ends and the other begins.  | Disabled |
|  `--hamming`, `-H` | Use Hamming distance instead of Levenshtein for variable barcodes. Only supported when allowing for one mismatch per barcode pattern. | Levenshtein |

The output directory can be set using `--output`, `-o`. All output files, including failed lines and statistics, are written to this output directory.

### Output: 
After demultiplex completes, it reports how well the sequencing reads were aligned to the patterns within the barcode scheme(s):
| Match type | Description |
| ---------- | ------------- |
| Perfect match | Reads whose barcodes match completely | 
| Moderate match | Reads that match within the allowed number of mismatches | 
| Mismatch | Reads that cannot be matched given the number of allowed mismatches | 

The output type and corresponding downstream step(s) depend on the barcode sequence encoding the feature modality.  
* **Generic barcode sequence**: Produces a *TSV-file* of reads split and aligned to the barcode scheme. The first column contains the read name, followed by the barcode patterns. This file is used directly as input for **count**.

* **Genomic sequence**, Extracts gene or transcript sequences into a *FASTQ-file* for alignment with **STAR** and subsequent annotation. The remaining read elements are aligned to the barcode scheme and saved in a *TSV-file* for use in downstream steps.

{% include alert.html type="info" title="Generic barcode sequences follow demultiplex → count, whereas genomic sequences follow demultiplex → STAR → annotate → count." %}

### STAR
For genomic sequence reads, where a barcode pattern encodes for a gene or transcript, **demultiplex** outputs a FASTQ file containing demultiplexed sequence reads. This *FASTQ* file is passed to the **STAR** aligner, which maps the sequencing reads to a reference genome. 

Before running STAR, the reference genome and GTF annotation files must be downloaded as described in [Reference Genome](getting-started#reference-genome). 

Key output files:

| File type | Description |
| --------- | ----------- |
| aligned.out.bam | Reads aligned to the reference genome
| TSV | Detected splice junctions and exon-intron boundaries

### Annotate
The output files generated by **STAR** contain information on the genomic coordinates are passed to **annotate**,to add te STAR annotation to the output of demultiplex

which integrates genomic coordinates with the barcodes encoding individual cells to produce the final *TSV* file.

Finally, both workflows output one *TSV* file per barcode scheme. These files contain all barcode-scheme-aligned sequencing reads, with the read name in the first column and positional barcode patterns in the subsequent columns.

## Count
The **count** submodule groups the aligned reads by single-cell and feature barcode. By default, identical UMI-tagged entries are collapsed to produce the final counts for each cell- feature combination. 

### Input & set up:
As input, **count** takes barcode-aligned reads *(TSV-file)* along with the barcode information, including barcode scheme, patterns and allowed mismatches. It also requires indices of the barcode patterns encoding the single-cell, feature, and UMI identities. Note that when defining these indexes, the **demultiplex** output *TSV* file lists the read name in de first column, followed by the positional barcode patterns. Also, when an identity is encoded by a combination of variable barcode patterns, a list of indices can be provided.

Counting begins at 0 and includes every pattern element defined within square brackets

Required parameters:

| Option | Description | Type |
| --------- | ----------- | ------ | 
| `--input`, `-i` | Barcode-aligned reads for features in single cells |TSV
| `--barcodeDir`, `-d` | Directory containing all barcode-related files, including barcode scheme, patterns, and mismatches | Directory
| `--FeatureIndex`, `-x` | Indices referring to the positions of the barcode patterns that encode feature identities | Comma-separated list
| `--singleCellIndices`, `-c` | Indices referring to the positions of the barcode patterns that encode the single-cell identities | Comma-separated list
| `--umiIndex`, `-u` | Indices referring to the positions of the barcode patterns that encode the UMI | Comma-separated list

Optionally, feature names and single-cell annotations, like experimental conditions, can be added by providing them in the same order as corresponding barcodes in the barcode pattern files. 

Optional parameters for annotation:

| Option | Description | File type | Default |
| ---------- | ------------- | ----------- | -------------- |
| `--featureNames`, `-a` | List of all feature names, in same order as the feature-barcodes in the barcode file | (.txt) | Nucleotide sequences
|  `--annotationFiles`, `-g` | List of paths to files, where every file contains a list of single-cell annotations | (.txt) | None
|  `--annotationIdxs`, `-y` | List of space separated indices used to annotate cells  | (.txt) | None
|  `--scIdAsString`, `-s` | Stores the single-cell ID as the sequence string instead of barcode ID | None | 1

By default, UMI collapsing is enabled using Hamming distance, allowing a single mismatch. Using the parameters below, you can disable UMI collapsing or change the distance measure and the number of allowed mismatches.

Optional parameters for UMI collapsing:

| Option | Description | Default |
| ---------- | ------------- | -------------- |
|  `--mismatches`, `-m` | Number of mismatches allowed during UMI collapsing | 1
|  `--hamming`, `-H` | Use Hamming instead of Levenshtein distance during UMI collapsing (allowing for insertions and deletions) | 0
|  `--umiThreshold`, `-f` | Threshold to filter UMIs before collapsing | 0
|  `--umiRemoval`, `-z` | UMIs are collapsed | 1

### Output
When **count** finishes, it creates a *LOG_COUNTS.tsv* file summarizing the total number of processed reads and the number of detected UMI mismatches. 

The key output are two count matrices *(TSV)*: 
- **UMI uncollapsed** : Reports the number of reads per UMI, providing insight into UMI amplification during the PCR experiment.
- **UMI collapsed**: All reads sharing the same UMI are collapsed to create the final count matrix for each unique single-cell and feature combination.

## ESGI
Instead of running **demultiplex** and **count** separately, you can execute them together using **ESGI**. 

To run **ESGI**, you need to provide an *initialization-file (.ini)* that specifies the required input, including paths to all barcode-related files, as well as indexing information. For indexing, counting starts at zero and includes every pattern element defined within square brackets. Optionally, you can also include feature names and additional annotation.

Example of ESGI Initialization-file:
```shell
# Input directory
forward="/path/to/forward_reads.fastq.gz"
# Reverse reads (optional)
reverse="/path/to/reverse_reads.fastq.gz"

# Output directory
output="/path/to/output"

# Paths to barcode scheme and mismatch settings
pattern="/path/to/barcode_pattern.txt"
mismatches="/path/to/mismatches.txt"

# Indexes for single-cell, feature, and UMI assignment
SC_ID=1,5
FEATURE_ID=3
UMI_ID=4

# Optional: assign feature names to barcode patterns corresponding to FEATURE_ID
FEATURE_NAMES="/path/to/feature_names.txt"

# Additional barcode files for naming annotated barcode patterns, like experimental conditions
ANNOTATION_IDs="/path/to/annotation_ID.txt"
ANNOTATION_NAMES="/path/to/annotation_names.txt"

threads=10
prefix=MYEXPERIMENT
```



### Versioning Options


| Parameter | Description | Values |
| --------- | ----------- | ------ | 
| versioning | This determines whether the versioning functionality is enabled | true \| false 
| version_menu | In addition to the version dropdown you can specify some text to display alongside it<br/> If you do not wish to use this then make this value an empty string | Release
| version_dir | This is the directory where all the alternative versions of the documentation will be contained beneath the _docs directory <br/>i.e. setting this to the value of Archive (Note: **No** trailing slash) will mean all alternative versions will be found in _docs/Archive/ | Archive
|  tocversion_dir | This is the directory inside your _data directory where Docsy Jekyll will find the alternative TOC files that are used for other versions of the documentation. You can configure this to be whatever name you like, just ensure your other TOCs can be found there! | versions 
| latest | From the list of versions this determines the one which relates to your base docs directory.<br/> Note the url for your docs will not contain a version identifier for this particular version| current \| v3.1
| allow_search | A boolean to indicate if we should allow searching over versions. If False, only the current is displayed in results | true \| false 
| search_version | A list of versions to include in the search. Current or latest will be tagged with the site.tag_color, the rest will be gray | current \| previous
| versions | A list of versions you wish to display in the version dropdown <br/> Note that there are several special keywords that can be used: current, main, alpha, beta, rc and pre, see below for more details  | current \| v1.0 \| v2.1.3

{% include alert.html type="warning" title="Note: The versions listed must have the same name as the folder name in which that version of the documentation resides otherwise jekyll it will not find it! The exception is the version 'current' which does not have a separate folder and is used to inform jekyll to utilize the base _docs folder." %}

### Version Keywords
There are six special keywords that can be used in the versions config.
* main
* alpha
* beta
* rc
* pre
* *current*

The value **main**, is used to display the latest development version of the documentation, a banner is used to denote that the code and documentation maybe unstable and that they are not guaranteed to be correct or up to date and could change at any time.

The value of **alpha**, **beta**, **rc** and **pre** are used to denote that the documentation is a pre-production version and as such a banner is displayed announcing that the code/documentation maybe unstable and the content in the pages are still work in progress so may not be completely up to date or correct.

If you wish to utilize this feature then you must create folders within your _docs/*Archive* directory with the names of those you wish to use and place your documentation inside it/them.

The value of **current** is used to align with the latest version of your documentation in your base _docs folder so there **doesn't** need to be a specific folder created and named 'current' to serve them. The name 'current', for the value of 'latest' in the previous example, can be changed to whatever you wish to name the current release but be aware that the version set in the _config.yml file also needs to match it! See the example below where we have called the latest release v3.2 rather than 'current' and we also have a version exactly named v3.2 too, as mentioned this version will point to your base _docs directory. 

```yml
version_params:
  version_menu: "Release"
  version_dir: Archive
  tocversion_dir: versions
  versioning: true
  latest: v3.2
  versions:
    - main
    - v3.2
    - beta
    - v3.1
```

### Configuring Version Alerts

Depending on the version keywords you may use Docsy Jekyll tries to determine if the documentation is historical (e.g. out of date) or is a upcoming release of information and will show one of the following banners to alert users to the fact that they are not reviewing the current version of the documentation. 

|![Historical Versions]({{ site.baseurl }}/assets/img/versionalertoutdated.png)|
|:--:| 
| *This banner is shown for historical versions of your documentation* |

|![Beta Versions]({{ site.baseurl }}/assets/img/versionalertbeta.png)|
|:--:| 
| *This banner is shown for pre-production versions of your documentation* |

|![Development Versions]({{ site.baseurl }}/assets/img/versionalertmain.png)|
|:--:| 
| *This banner is shown for development versions of your documentation* |

These banners can be customised to have the wording or imagery you require by altering the HTML in the _includes/versionalert.html file.

## Table of Contents Handling (TOC)

As with the documents, your TOC contents also need to be updated so they correctly point to the document versions you are viewing.

You will need to update the toc file as the structure of your documentation/toc changes over time but as we have taken a snapshot of our documentation so we need to do the same thing for our toc so it reflects the document structure at that point in time too. You will see that for this site we have the versions of Current and Previous and you will see that the Current version has all the information about versioning which didn't exist in the prior version.

To enable the TOC to change based on the version, we have to create a new toc file and name it {version}-toc.yml and it must be placed in a subdirectory of _data whose name is specified by the tocversion_dir parameter in the _config.yml file (for this site it can be found in the _data/versions subdirectory). We have created previous-toc.yml in the _data/versions directory that has the TOC structure for the prior version where we did not have the versioning.md file present. 


To create this version of the toc file file we can use the following commands
```shell
# Our last release is version "Previous" so we want to create a TOC for that inside our versions directory
mkdir -p _data/versions
# copy the existing TOC file to our version subdirectory and name it as {version}-toc.yml
cp  _data/toc.yml _data/versions/previous-toc.yml
```


{% include alert.html type="info" title="Note: The toc.yml file will always be our current table of contents pointing to the most recent version of the documentation in the base of our _docs folder. When adding additional items to your toc.yml file you do not need to worry about the version identifier as that is taken care of by Docsy Jekyll when viewing alternative versions, continue to have the urls pointing to the files in the base docs directory." %}

Here is what our current toc.yml file in our _data directory looks like, the previous-toc.yml file will look identical apart from omitting the 'Versioning' title and link which was added for this release, and of course will live in the versions subdirectory of _data as well.

```yml
title: Documentation
  url: docs
  links:
    - title: "Getting Started"
      url: "docs/getting-started"
      children:
        - title: Features
          url: "docs/getting-started#getting-started"
        - title: Development
          url: "docs/getting-started#development"
        - title: Customization
          url: "docs/getting-started#customization"
    - title: Versioning
      url: "docs/versioning"
    - title: "About"
      url: "about"
    - title: "News"
      url: "news"
- title: "Extras"
  url: "docs/extras"
  links:
    - title: Quizzes
      url: "docs/extras/example-quiz"
    - title: Tags Page
      url: "tags"
```

Now not every version of your documentation will have a change in structure and therefore require an update to toc.yml. To remove this overhead a mapping file is used to tell Docsy Jekyll which TOC it should use for a particular version,
the benefit of this is that if the structure of your site does not change then you do not need to create an additional toc.yml file.

You will have a toc-mapping.yml file in your _data directory and the contents will contain the code below to map a particular version with a particular TOC file to use. You need to update this file for each version you create even if you do not need to create a new TOC file.

```yml
# This file can be used to explicitly map a release to a specific table-of-contents
# (TOC). You'll want to use this after any revamps to information architecture, to ensure
# that the navigation for older versions still works.

Current: toc
Previous: previous-toc
```
**You do not need to prefix the name of the new TOC filename with the folder (in our case versions) as Docsy Jekyll will retrieve that from the _config.yml file and add that to the path.**  

{% include alert.html type="info" title="Note: If a version cannot be found in this mapping file then the standard toc.yml in the _data directory will be used instead." %}

Because we added the versioning.md markdown file to this site we needed a new toc so that this new content could be accessed from the TOC menu. If we had only made an update to the content of the files that already existed, then we could have pointed Previous to toc in the yml above to use the same toc.yml file.

## Versioning and Search

As you will have seen in the [Versioning Options](#versioning-options) you have the ability to configure Docsy Jekyll to perform it's searches over the entire site's documentation, or only certain versions which you can specify. If you choose not to enable it in _config.yml (allow_search = false) then the searches that get performed will only be over the most current version of the documentation in your base _docs folder. If you prefer to have this enabled then you will need to provide a list of search versions (search_version in _config.yml) that you wish to be displayed if a match is found. The following image shows what your users would be shown if you have this enabled. You can see that the search was for the term 'start' and we have several results returned but two both for the 'Getting Started' topic. The reason for this is due to a result being returned within the current version as well as a result being returned for the previous version as denoted by the badges of 'Current' and 'Previous' below the titles. 


| *Searching with Versioning enabled*|
|:--:|
|![Search with Versions]({{ site.baseurl }}/assets/img/docsy-jekyll-version-search.png)


## Issues with Permalinks

Please do not use permalinks in your archive documentation Front Matter if you wish to use versioning as this will break the versioning system. If you have permalinks in your current documentation you will need to remove them when you place the documents into a subdirectory so that Jekyll can serve them correctly. 

{% include alert.html type="warning" title="Using permalinks in your archive documentation Front Matter will cause issues with versioning as this effectively 'hardcodes' the url of that page and will break the versioning links. Please be aware!" %}

## 404 Errors

There will be times where users will navigate around the documentation and could end up pointing to an invalid page that doesn't exist. For instance while you are reading this page on the **Current** version if you were to change this to the Previous version, via the dropdown, then you will get a page not found as this versioning page did not exist in the prior version. We have created a 404.md page in the base of the site so that the user still gets to navigate even though the page was not found. You can update this page with your own information should you so wish.

{% include alert.html type="info" title="When pages are not found and the 404 page is displayed the TOC menus default back to links pointing to the current release." %}
